okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -113,13 +113,23 @@ class ExchangeFinder(
           connectionRetryEnabled = connectionRetryEnabled
       )
 
-      // Confirm that the connection is good. If it isn't, take it out of the pool and start again.
-      if (!candidate.isHealthy(doExtensiveHealthChecks)) {
-        candidate.noNewExchanges()
-        continue
+      // Confirm that the connection is good.
+      if (candidate.isHealthy(doExtensiveHealthChecks)) {
+        return candidate
       }
 
-      return candidate
+      // If it isn't, take it out of the pool.
+      candidate.noNewExchanges()
+
+      // Make sure we have some routes left to try. One example where we may exhaust all the routes
+      // would happen if we made a new connection and it immediately is detected as unhealthy.
+      synchronized(connectionPool) {
+        val routesLeft = routeSelection?.hasNext() ?: true
```

Comment:
It seems like line 108 must have made routeSelection non null, so can we change to !!

Reply:
These will be `null` if the call had a previous connection we could use. Here's a test that demonstrates it:

```java
    server.enqueue(new MockResponse().setResponseCode(401));
    server.enqueue(new MockResponse().setResponseCode(200));
    AtomicReference<Connection> ref = new AtomicReference<>();
    EventListener listener = new EventListener() {
      @Override public void connectionAcquired(Call call, Connection connection) {
        ref.set(connection);
      }
    };

    client = client.newBuilder()
        .eventListener(listener)
        .authenticator((route, response) -> {
          Util.closeQuietly(ref.get().socket());
          return response.request();
        })
        .build();

    executeSynchronously("/").assertCode(200);
```